File: 410-split-array-largest-sum/split-array-largest-sum.java
class Solution {
    public boolean cntK(int k , int  mid , int[] nums){
        int cnt = 1 , s = 0 ;

        for(int i = 0 ; i < nums.length ; i++){
            if(s + nums[i] > mid){
                cnt++ ;
                s = nums[i] ;
                if(cnt > k)return false ;
            }else{
                s += nums[i] ;
            }
        }
        return true ;
    }
    public int splitArray(int[] nums, int k) {

        int low = Arrays.stream(nums).max().getAsInt() ;
        int high = Arrays.stream(nums).sum() ;

        while(low <= high){
            int mid = low  + (high - low) / 2 ;

            if(cntK(k , mid , nums)){
                high = mid - 1 ;
            }else{
                low = mid + 1 ;
            }
        }
        return low ;
    }
}

File: 410-split-array-largest-sum/split-array-largest-sum.cpp
#define all(v) begin(v),end(v)
class Solution {
private:
    bool cntK(int mid , vector<int>& nums , int k){
        
        int s = 0 ;
        int cnt = 1 ;

        for(auto &num : nums){
            if(s + num > mid){
                cnt++ ;
                s = num ;
                if(cnt > k)return false ;
            }else{
                s += num ;
            }
            
        }
        return true ;
    }
public:
    int splitArray(vector<int>& nums, int k) {

        int low = *max_element(all(nums)) ;
        int high = accumulate(all(nums) , 0) ;

        while(low <= high){
            int mid = high +(low - high) / 2 ;
            if(cntK(mid , nums , k)){
                high = mid - 1 ;
            }else{
                low = mid + 1 ;
            }
        }
        return low ;
    }
};